docs: new tutorial "validate asyncapi documents using studio" (#1022)
asyncapi/website: pages/docs/tutorials/validate-documents.md

---

> title: "Studio AsyncAPI document validation"

Reviewer: should not titles follow some single rule, like call to action or something else

wouldn't it sound better as `Validate AsyncAPI document with Studio`?

Author: We thought of a sentence but it doesn't look good in the side navigation bar. We'll be getting a long sentence in two parts

Reviewer: ok, but on my screen `Studio AsyncAPI document validation` is also split

<img width="315" alt="Screenshot 2022-11-22 at 14 46 25" src="https://user-images.githubusercontent.com/6995927/203329752-71fdff47-920a-4164-abc0-e544a849a280.png">

maybe we should consider making left nav wider. I kinda do not like that we need to adjust titles to fit screen, you know, it is like writing content for the browser not for the human 😄

Author: okay then I'll make it as Validate AsyncAPI document with Studio

> description: This tutorial will teach us how to validate AsyncAPI documents using the AsyncAPI Studio tool.

Reviewer: ```suggestion
description: This tutorial will teach you how to validate AsyncAPI documents using the AsyncAPI Studio tool.
```

weight: 120
---

# Introduction
This tutorial will teach you how to validate AsyncAPI documents using the [AsyncAPI Studio tool](https://studio.asyncapi.com/).

> We will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).

Reviewer: ```suggestion
You will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).
```

# Background context
An AsyncAPI document is a file that defines and annotates the different components of a specific Event-Driven API. The format of the file must be JSON or YAML. You can use this document to generate both documentation and code.

The AsyncAPI Studio tool allows you to develop an AsyncAPI document, validate it, preview it, convert it to the latest version, and visualize event flows.

# Copy invalid AsyncAPI document
Let's pretend we have an invalid AsyncAPI document.

1. Open [Studio](https://studio.asyncapi.com/).

2. Navigate to the left side bar and click on **new file**.

3. Copy and paste the below invalid AsyncAPI document:

> ```yaml

Reviewer: example should be based on the same file as the other tutorials.

also I think we should be consistent with other tutorials.

so say: "yo, if you go through this tutorial without performing previous steps, then get an asyncapi file with `asyncapi new --example=tutorial.yml --no-tty`. Paste it in Studio, all works perfect, documentation is rendered. Not let's have some fun to see how errors with your file are displayed. Go and break this and that in your document. You see not docs are not rendered and in diagnostics you see some more details"

otherwise we do not have tut 1 -> tut 2 -> tut 3 flow

Reviewer: also `asyncapi: '2.7.0'` is risky, we might reach it april next year 😄 

maybe do `asyncapi: '1.0.0'`

Author: I'm not sure where to paste `asyncapi new --example=tutorial.yml --no-tty` in studio? I don't see any terminal or anything like that which can process this command @derberg

Author: ![image](https://user-images.githubusercontent.com/74637789/203547408-d3f5bc7c-2b44-470b-984a-0a71f5206f81.png)
I see this in studio too, how useful will that be?

Reviewer: this is something people would have to generate on local first, and copy/paste

although, we could also enable the same tutorial.yml in the Studio, when one clicks on 
<img width="240" alt="Screenshot 2022-11-24 at 12 04 39" src="https://user-images.githubusercontent.com/6995927/203768894-15a4d961-fb06-402b-82cb-6d7b56407486.png">

🤔

Reviewer 2: Do you wanna have invalid AsyncAPI document as template in the Studio? 😄

Author: oh actually if that helps but that will be used only for this tutorial "educational purposes only" right?

Reviewer 2: It's not for me to decide, but we can create templates only for tutorials, it's not a problem.

Reviewer: > It's not for me to decide

@magicmatatjahu kinda is as you are one of Studio maintainers 😆 

but yeah, that would be marked as Tutorial with description that it is for education with docs only

Reviewer 2: Ok, should I create PR or @starlightknown? I can help.

Reviewer: just folks remember, in studio we need current yaml 😄 

Just flow of this tutorial must indicate that user selects valid template and then just breaks the document in studio to simulate errors

asyncapi: '2.7.0'
info:
  title: Account Service
  version: 1.0.0
  
channels:
  user/signedup:
    subscribe:
      message:
        $ref: 'UsersSignedUp'
components:
  messages:
    UserSignedUp:
      payload:
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user
 ``` 

> # Troubleshoot Studio console errors

Reviewer: what console errors you mean?

Reviewer: studio just changed a bit, have a look

Author: oh it's diagnostics now, fixed

Let's fix the errors one by one until we end up with a valid AsyncAPI document.

> 1. In Studio, open the **visualizer** found on the left-side menu.

Reviewer: why do I need to open visualizer to troubleshoot errors?

Author: fixed it

5. Read the next studio console error: `Error downloading https://studio.asyncapi.com/UserSignedUp HTTP ERROR 404`.

6. Fix the `$ref` by changing it to: `'#/components/messages/UserSignedUp'`.

```yaml
channels:
  user/signedup:
    subscribe:
      message:
        $ref: 'UserSignedUp'
```

This allows the consumer applications to subscribe to the _user/signedup_ channel and receive `userSignUp` messages. 

<Remember>
The <b>channels</b> section is used to describe the event names your API will be publishing and/or subscribing to.
</Remember>

7. Congratulations! You identified and fixed all the errors, and now have a valid AsyncAPI document.

# Summary
This tutorial taught us how to validate an AsyncAPI document using the AsyncAPI Studio tool. We also learned to troubleshoot an invalid AsyncAPI document by following the console error message directions. In doing so, we learned how to identify `REQUIRED` properties in all AsyncAPI documents.

# Next steps
Now that you've completed this tutorial, go ahead to learn how to [validate messages/events]() which you will be sending to your application.
---

<DocsButton
  suggestions={[
    {
      href: '/docs/tutorials/generate-code',

>       title: 'Generate Code',

Reviewer 3: ```suggestion
      title: 'Generate code',
```

      type:'back',
    },
    {

>       href: '/docs/tutorials/validate-events',
>       title: 'Messages/events validation',

Reviewer 3: ```suggestion
      href: '/docs/tutorials/message-validation',
      title: 'Message validation',
```

Heyo, can we [please update our file names](https://github.com/asyncapi/website/pull/1061/files) to this? we shouldn't use "events" since we already decided to go with the word `messages`. 😄✌🏽

---
title: "Validate AsyncAPI document with Studio"
description: This tutorial will teach you how to validate AsyncAPI documents using the AsyncAPI Studio tool.
weight: 120
---

## Introduction
This tutorial will teach you how to validate AsyncAPI documents using the [AsyncAPI Studio tool](https://studio.asyncapi.com/).

> You will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).

Reviewer 4: ```suggestion
You will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/latest#A2SObject).
```

@alequetzalli @derberg shouldn't the link for the specification be directed to the latest specification, instead of just pointing to v2.5.0?

Reviewer 3: yes! thank you for that idea

## Background context
An AsyncAPI document is a file that defines and annotates the different components of a specific Event-Driven API. The format of the file must be JSON or YAML. You can use this document to generate both documentation and code.

The AsyncAPI Studio tool allows you to develop an AsyncAPI document, validate it, preview it, convert it to the latest version, and visualize event flows.

If you are following tutorial without doing the previous tutorial [Create AsyncAPI document]() then quickly generate an AsyncAPI file using `asyncapi new --example=tutorial.yml --no-tty`. You can see that the documentation is now rendered and you don't see any errors which means your file is valid. Now let's experiment with an invalid file to see how errors are displayed and how to make that file valid again.

## Copy invalid AsyncAPI document
Let's pretend we have an invalid AsyncAPI document.

1. Open [Studio](https://studio.asyncapi.com/).

2. Copy and paste the below invalid AsyncAPI document:

```yaml

> {`asyncapi: '1.0.0'

Reviewer: hmm, where did you take this from, it refers to ancient asyncapi 1.0.0 😄 
also please remove 
```
{`
```
and 
```
`}
```

 ``` 
## Troubleshoot Studio errors
Let's fix the errors one by one until we end up with a valid AsyncAPI document.
 
1. You can see the error message on the screen: `Empty or invalid document. Please fix errors/define AsyncAPI document.`

2. Open diagnostics, you can see more information related to your errors.

3. Fix the incorrect AsyncAPI specification number to `2.5.0`.
 
```yaml
asyncapi: '2.5.0'
info:
  title: Account Service
  version: 1.0.0
  ```
<Remember>
Notice how <b>description</b> property is missing; that doesn't make the AsyncAPI document invalid, but it's always better to include.
</Remember>

4. Read the next error: `must be number`. Fix the `minimum` by changing it to: `0`.

```yaml
          properties:
            id:
              type: integer
              minimum: 0
``` 
5. You see three errors:
- must be equal to one of the allowed values
- must be array
- must match a schema in `anyOf`

`anyOf` means it should match any one the above schemas then it is valid.

Now let's fix this error by changing the type to `string`

```yaml
            sentAt:
              type: string
              format: date-time

>               description: Date and time when the message was sent.`}

Reviewer: ```suggestion
              description: Date and time when the message was sent.
```

```

6. Congratulations! You identified and fixed all the errors, and now have a valid AsyncAPI document.

## Summary
This tutorial taught us how to validate an AsyncAPI document using the AsyncAPI Studio tool. We also learned to troubleshoot an invalid AsyncAPI document by following the error message directions in diagnostics. In doing so, we learned how to identify `REQUIRED` properties in all AsyncAPI documents.

## Next steps

> Now that you've completed this tutorial, go ahead to learn how to [validate messages/events]() which you will be sending to your application.

Reviewer: next step do not rever to the same doc as the buttons below, which is strange

also look at https://deploy-preview-1022--asyncapi-website.netlify.app/docs/tutorials/validate-documents#now-that-youve-completed-this-tutorial-go-ahead-to-learn-how-to-validate-messagesevents-which-you-will-be-sending-to-your-application

strange stuff with section, and to be honest I don't  know why 🤔 
try maybe `---` must have an empty line before

Author: fixed

description: This tutorial will teach you how to validate AsyncAPI documents using the AsyncAPI Studio tool.
weight: 120
---

## Introduction
This tutorial will teach you how to validate AsyncAPI documents using the [AsyncAPI Studio tool](https://studio.asyncapi.com/).

You will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).

## Background context
An AsyncAPI document is a file that defines and annotates the different components of a specific Event-Driven API. The format of the file must be JSON or YAML. You can use this document to generate both documentation and code.

The AsyncAPI Studio tool allows you to develop an AsyncAPI document, validate it, preview it, convert it to the latest version, and visualize event flows.

<Remember>

If you did not follow the previous tutorial and do not have an `asyncapi.yaml` file ready, then generate one using `asyncapi new --example=tutorial.yml --no-tty` command.

> </Remember>

Reviewer: btw, I saw this in at least 2 tutorials already. Might make sense to turn it into fragment too

Reviewer 3: let's create new fragments for items longer than just 1 sentence tho

---
title: "Studio AsyncAPI document validation"
description: This tutorial will teach us how to validate AsyncAPI documents using the AsyncAPI Studio tool.
weight: 120
---

# Introduction
This tutorial will teach you how to validate AsyncAPI documents using the [AsyncAPI Studio tool](https://studio.asyncapi.com/).

We will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).

> 

Reviewer 3: @starlightknown Hey, you're missing the `Installation Guide` section. Please add it and tag us to let us know when it's ready to review again.

Author: Hey we discussed about this in slack and you told me that we don't need installation guide. This is done in studio so we don't need anything to install locally. Let me know your thoughts

Reviewer 3: Oooh you're right, we talked about this 😂. Sorry, my bad 🤪

# Background context
An AsyncAPI document is a file that defines and annotates the different components of a specific Event-Driven API. The format of the file must be JSON or YAML. You can use this document to generate both documentation and code.

The AsyncAPI Studio tool allows you to develop an AsyncAPI document, validate it, preview it, convert it to the latest version, and visualize event flows.

# Copy invalid AsyncAPI document
Let's pretend we have an invalid AsyncAPI document.

1. Open [Studio](https://studio.asyncapi.com/).

2. Navigate to the left side bar and click on **new file**.

3. Copy and paste the below invalid AsyncAPI document:

> ```

Reviewer 2: ```suggestion
```yaml
```

asyncapi: '2.7.0'
info:
  title: Account Service
  version: 2.0.0
  
channels:
  user/signedup:
    subscribe:
      message:

>       $ref: 'UsersSignedUp'

Reviewer 2: ```suggestion
        $ref: 'UsersSignedUp'
```

components:
  messages:
    UserSignedUp:
      payload:
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user
 ``` 

# Troubleshoot Studio console errors
Let's fix the errors one by one until we end up with a valid AsyncAPI document.

1. In Studio, open the **visualizer** found on the left-side menu.
 
2. Read the error message on the visualizer preview screen: `Empty or invalid document. Please fix errors/define AsyncAPI document.`

3. Open the console box that says `PROBLEMS`, so you can read all console errors.

> 4. Fix the incorrect AsyncAPI specification number to `2.4.0` and the `version`number to `1.0.0`.

Reviewer 2: FYI Version `2.0.0` defined in `info.version` doesn't produce any error, because it can have any value. It should follow by [semver](https://semver.org/) format, but as I see we don't validate that 😆 I will create issue for it.

Author: Shall I keep this step in guide then? Or remove it

Reviewer 2: We can remove info about `version` field, yes :) Please also change `version` field to the `1.0.0` value in previous yamls, ok?

Author: done, thank you

>  ```

Reviewer 2: ```suggestion
```yaml
```

>  asyncapi: '2.4.0'

Reviewer 2: ```suggestion
asyncapi: '2.4.0'
```

info:
  title: Account Service
  version: 1.0.0
  ```

<Remember>
Notice how <b>description</b> property is missing; that doesn't make the AsyncAPI document invalid, but it's always better to include.
</Remember>

5. Read the next studio console error: `Error downloading https://studio.asyncapi.com/UserSignedUp HTTP ERROR 404`.

6. Fix the `$ref` by changing it to: `'#/components/messages/UserSignedUp'`.

> ```

Reviewer 2: ```suggestion
```yaml
```

channels:
  user/signedup:
    subscribe:
      message:

>       $ref: 'UserSignedUp'

Reviewer 2: YAML is based on indentations, so if you wanna define `$ref` in `message` you need to make indentation in this block, otherwise YAML understand it that `$ref` is defined in `subscribe` block, not `message`

```suggestion
        $ref: 'UserSignedUp'
```
